cpp/arcticdb/storage/failure_simulation.hpp

```diff
@@ -7,56 +7,148 @@
-    READ
+    READ,
+    ITERATE,
+    DELETE,
 };
 
 static const char* failure_names[] = {
         "WRITE",
-        "READ"
+        "READ",
+        "ITERATE",
+        "DELETE",
 };
 
-struct FailureCategory {
-    double prob_;
-    bool thrown_;
+/** Function holder with a description. */
+struct FailureAction {
+    using FunctionWrapper = folly::Function<void(FailureType)>;
+
+    const char* description_;
+    std::string description_str_; // When we need to manage the life-cycle of the description_
+    FunctionWrapper::SharedProxy proxy_;
+
+    FailureAction(const char* description, FunctionWrapper::SharedProxy proxy) :
+            description_(description), proxy_(std::move(proxy)) {}
+
+    FailureAction(std::string description, FunctionWrapper::SharedProxy proxy) :
+            description_str_(std::move(description)), proxy_(std::move(proxy)) {
+        description_ = description_str_.c_str();
+    }
 
-    explicit FailureCategory(double prob) :  prob_(prob), thrown_(false) {}
+    template<typename Desc, typename Func>
+    FailureAction(Desc description, Func func):
+        FailureAction(description, FunctionWrapper{func}.asSharedProxy()) {}
+
+    inline void operator()(FailureType type) const {
+        proxy_(type);
+    }
+};
+
+inline std::ostream& operator<<(std::ostream& out, const FailureAction& action) {
+    return out << action.description_;
+}
+
+namespace action_factories { // To allow `using namespace`
+    static inline FailureAction no_op("no_op", [](FailureType){});
+
+    /** Raises the given exception with the given probability. */
+    template<class Exception = StorageException>
+    static inline FailureAction fault(double probability = 1.0) {
+        util::check_arg(probability >= 0, "Bad probability: {}", probability);
+
+        static FailureAction do_fault("raise", [](FailureType failure_type) {
+            throw Exception(fmt::format("Simulating {} storage failure", failure_type));
+        });
+
+        if (probability >= 1.0) {
+            return do_fault;
+        } else {
+            return {fmt::format("fault({})", probability),
+                    [prob=probability](FailureType failure_type) {
+                        thread_local std::once_flag flag;
+                        std::call_once(flag, [seed = uint64_t(&failure_type)]() { init_random(seed); });
+                        if (random_probability() < prob) {
+                            do_fault(failure_type);
+                        }
+                    }};
+        }
+    }
+}
+
+/** Independent state for each FailureType. Thread-safe except for the c'tors. */
+class FailureTypeState {
+public:
+    using ActionSequence = std::vector<FailureAction>;
+    static_assert(std::is_copy_assignable_v<ActionSequence>);
+
+private:
+    friend class StorageFailureSimulator;
+
+    const ActionSequence sequence_;
+    std::atomic<size_t> cursor_ {0}; // Index into sequence
+
+public:
+    FailureTypeState(ActionSequence sequence) :
```

Comment:
Please mark single arg constructors explicit.

Reply:
As this class cannot be moved, the map (entry) containing this can only be constructed using the `piecewise_construct` trick below, so this can't be explicit.

Reply:
I don't get why can't this be explicit. Is it a compilation error, I haven't heard of any rules forbidding the explicit specifier.